multimc: init

Forkk opened the pull request:
###### Motivation for this change

This package provides MultiMC, a popular alternative Minecraft launcher,
packaged in a way that is approved and supported by upstream. The package
includes a script which will download the official binaries from upstream and
install to `~/.local/share/multimc`, as well as a `.desktop` file, and an FHS
environment which mimics the environment MultiMC expects.

Obviously this is not exactly in line with the ideal way to package software
for NixOS, but this particular software relies on multiple 3rd parties who
often introduce breaking changes with little to no notice, and as such,
upstream has requested that it be packaged in a way that allows it to
self-update, so that these breaking changes can be resolved quickly without
having to wait on distro package updates.

This package also provides an argument for managing which versions of Java are
made available to MultiMC. This is done by adding them to `/opt/jdks` in the
FHS environment, which is one of the paths MultiMC searches for Java installs.
This allows MultiMC to have a consistent path to the JRE that doesn't change
when NixOS updates Java. It also provides an argument to configure extra JREs
for users who want them.

The FHS chroot environment includes all of the necessary libraries for vanilla
Minecraft, as well as most mods, but it is possible for some mods to introduce
their own native libraries with additional dependencies, so this package also
provides an argument for adding extra packages to the environment.


###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [X] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Forkk committed:
multimc: init

LunNova commented:
Having it self-update doesn't feel nice to me, you're missing most of the point of nix and you can't test against our deps. Since there is a fork which doesn't insist on this, what is the benefit of having a multimc package?

This could be provided as an unofficial flake/overlay somewhere instead. I don't think it's really getting any of the usual benefits of being in nixpkgs as nothing can be tested and you can't verify it will really work with the provided deps.

If this does go in, I think it should get a new attrpath `multimc-bin` in keeping with other upstream binary packages and keep the notice about the alternative polymc so that users know they can use a package that is packaged normally.

LunNova commented:
cc @muscaln @starcraft66

Forkk commented:
> Having it self-update doesn't feel nice to me, you're missing most of the point of nix and you can't test against our deps. Since there is a fork which doesn't insist on this, what is the benefit of having a multimc package?

As a NixOS user, I understand the concern with having packages that self-update, but as mentioned in my motivation for this change, this software is prone to breakage by 3rd parties, and would see little benefit from a lot of what NixOS provides for most packages. For example, any version of MultiMC over a few months old is likely to have serious issues, or not work at all, so it would be pointless to pin to a specific version.

The automatic updates allow upstream to resolve these issues quickly without waiting on packagers to update.

The benefit of having this package is that in our support channels, we have a package that we can point NixOS users at which operates in a fairly "normal" way that people who have volunteered to support this software already understand, whether they use NixOS or not. It also provides Java at consistent paths, unlike what you would get by using `steam-run` for example.


> This could be provided as an unofficial flake/overlay somewhere instead. I don't think it's really getting any of the usual benefits of being in nixpkgs as nothing can be tested and you can't verify it will really work with the provided deps.

I will certainly do this if this can't be merged here, but I think it would be easier on everyone to just have this available without users having to add an overlay.

> If this does go in, I think it should get a new attrpath `multimc-bin` in keeping with other upstream binary packages and keep the notice about the alternative polymc so that users know they can use a package that is packaged normally.

I'm certainly open to doing this. It would be more consistent. I'll wait and see what other people have to say though.

MagicRB commented:
I for one definitely want to remain on the source built and nixpkgs pinned version of multimc. So `multimc` and `multimc-bin` are a good compromise imo.

Forkk commented:
> I for one definitely want to remain on the source built and nixpkgs pinned version of multimc. So `multimc` and `multimc-bin` are a good compromise imo.

The original source-built `multimc` package was removed a while ago in favor of `polymc`. I would recommend not using it anymore, since it is already fairly far behind the latest version and doesn't seem to be getting updated any more due to its replacement.

starcraft66 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31052,6 +31052,8 @@ with pkgs;
 
   mrrescue = callPackage ../games/mrrescue { };
 
+  multimc = callPackage ../games/multimc { };
```
```suggestion
  multimc-bin = callPackage ../games/multimc { };
```

Since this is not being built from source for a reason other than the build from source being too cumbersome or its infrastructure being missing from nixpkgs, this package should clearly hold the `-bin` suffix, just like `firefox-bin` etc.

Forkk replied:
Package has been renamed to `multimc-bin`.

starcraft66 commented on pkgs/games/multimc/run.sh:
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
This kind of impurity is disgusting, worse than a fixed-output derivation. I don't think we should be allowing something like this in nixpkgs. If you want to control the version of the software you are distributing to users, please provide an upstream flake for others to use instead.

Forkk replied:
I think disgusting is a bit harsh here. You'll find other self-updating software in nixpkgs already, like Steam. I don't think it's reasonable to expect all software to conform perfectly, and one of NixOS's advantages is that it can still find a place for that software to fit in.

Ultimately whether to include this or not should be a question of practicality, and not some ideal of purity. The question is whether we inconvenience people by making them add an unofficial overlay, or give them the convenience of an official package in nixpkgs. I would prefer the latter.

starcraft66 replied:
Steam is self-updating in nixpkgs because valve only distributes a few latest versions of the client iirc. Steam breaks every once in a while as it starts depending on new libraries as a result of that. Discord is a good example of self-updating software we have to "work" around but can still provide pinned revisions for because upstream provides a history of old versions.

Forkk replied:
> Steam is self-updating in nixpkgs because valve only distributes a few latest versions of the client iirc.

And it is often the case that the last few versions of MultiMC are the only ones that actually still work. I don't see how that situation is any different just because the previous broken versions are still distributed. Who would want to use software that is no longer functional?

As for Discord, having to wait for packagers to update it whenever they decide to "break" the previous version is a major pain point for me and lots of other people and I would appreciate if that self updated too.

I expect my NixOS system to continue functioning if I change nothing, and these packages that end up with broken old versions do not allow that.

jyooru replied:
> > Steam is self-updating in nixpkgs because valve only distributes a few latest versions of the client iirc.
>
> And it is often the case that the last few versions of MultiMC are the only ones that actually still work. I don't see how that situation is any different just because the previous broken versions are still distributed.

The situation is different because Steam is not FOSS. MultiMC goes against the ideas of FOSS software by imposing restrictions like this on package repositories.

Restrictions on branding material is a little annoying, but understandable.

Trying to add an unreproducible FOSS package to a package repository that focuses on being reproducible, however? That sure is something else.

Forkk replied:
> The situation is different because Steam is not FOSS. MultiMC goes against the ideas of FOSS software by imposing restrictions like this on package repositories.
> 
> Restrictions on branding material is a little annoying, but understandable.

All of the packaging restrictions imposed *are* branding restrictions. If you remove the branding, you can package the software however you like, as PolyMC has done. Maybe this goes against your ideas of FOSS software, but it's in line with the license, which does not grant anyone the rights to use trademarks, only the code. In fact, Firefox, which I think most people would agree is free software, has similar requirements regarding branding.

> Trying to add an unreproducible FOSS package to a package repository that focuses on being reproducible, however? That sure is something else.

To me, one of the great benefits of nixpkgs is that a lot of software that is stuck in the AUR on Arch, or in some overlay on Gentoo, is right here in the official repository. That's why I've made this PR instead of just making a 3rd party flake. If this is not welcome here, then I will just do it the less convenient way with a flake.

starcraft66 replied:
I don't see any branding issues here, only that you're not actually packaging multimc in a reproducible way. The download of the binary should be locked to a specific revision. I imagine discussion around this topic would go in the same direction regardless of the linux distribution's repos you'd be trying to introduce such a package into.

mohe2015 replied:
> The download of the binary should be locked to a specific revision.

I agree. In particular there is a huge difference between e.g. firefox-bin and steam-bin which both don't modify the binary itself. Sure steam downloads and runs games if you instruct it to but it doesn't just run arbitrary code out of nowhere.

This in turn makes this pretty useless as then you could also build from source.

Forkk committed:
multimc -> multimc-bin

babbaj commented on nixos/doc/manual/from_md/release-notes/rl-2205.section.xml:
```diff
@@ -199,6 +199,12 @@
+      <listitem>
+        <para>
+          <link xlink:href="https://github.com/PolyMC/PolyMC">polymc</link>,
+          a free, open source launcher for Minecraft.
```
```suggestion
          an open source launcher for Minecraft.
```

Forkk replied:
Not only is this a childish thing to do, you've done it to the wrong bit of software. Nice job.

babbaj replied:
My bad, I just added a suggestion to the multimc description

babbaj commented on pkgs/games/multimc/default.nix:
```diff
@@ -0,0 +1,97 @@
+  '';
+
+  meta = with lib; {
+    description = "Free, open source launcher and instance manager for Minecraft";
```
```suggestion
    description = "Open source launcher and instance manager for Minecraft";
 ```

Forkk replied:
I don't appreciate this sort of poor attempt at trolling. We're here to discuss the merits of this PR, and what actual practical changes should be made to it, not to take cheap shots at people we disagree with on software licensing matters.

babbaj commented on pkgs/games/multimc/default.nix:
```diff
@@ -0,0 +1,97 @@
+# This works better than pointing MultiMC at a JVM inside the nix store, as
+# doing that may result in said JVM disappearing when the user collects
+# garbage.
+jvmSymlinks = stdenv.mkDerivation rec {
```
This can be implemented with runCommand, similar to how symlinkJoin works

Forkk replied:
Thanks, I will definitely look into that.

kenzieeeeee commented:
MultiMC is maintained by a Microsoft employee who has falsely smeared people who repackaged it as rapists, relicensed the code against contributors wishes, and stripped parts of the public codebase to make it harder to build and package independently. There is little reason to use it over PolyMC and it seems MultiMC is going down the path of ['embrace, extend, extinguish'](https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish).

These reasons are all problematic on their own, but the smearing part is especially egregious. Not only is the pull request here likely made by an alt of peterix, accepting it would also be normalizing and endorsing rape.

TacoGS commented:
Forkk was the original developer of MultiMC who did a lot of the leg work before others joined in. I would advise if you're going to turn a pull request into a battleground at least know who you're engaging.

Forkk commented:
> Not only is the pull request here likely made by an alt of peterix

You are severely misinformed. I am the original developer of MultiMC, a project I started 10 years ago. You can easily verify that by just looking at my GitHub profile. Suggesting I'm an alt of peterix is probably the funniest thing I've heard all day though, so thanks for that.


> MultiMC is maintained by a Microsoft employee who has falsely smeared people who repackaged it as rapists, relicensed the code against contributors wishes, and stripped parts of the public codebase to make it harder to build and package independently. There is little reason to use it over PolyMC and it seems MultiMC is going down the path of ['embrace, extend, extinguish'](https://en.wikipedia.org/wiki/Embrace,_extend,_and_extinguish).

Though you are completely mis-stating what peterix actually said on that flathub PR, I agree that it was indefensible, and I won't excuse it. However, the rest of this is pure nonsense. The codebase was "stripped" of branding to make it easier for forks to use different branding, and we are required by microsoft to keep the MSA ID needed for authentication as hidden as is practical. You can still register your own ID and build the software yourself if you like. Our only ask is that you don't distribute these builds to other people under the name "MultiMC," and that you don't direct users to our support channels, which is totally reasonable IMO.

JJJollyjim commented:
I would much rather we go down the path of building our own version.

Forkk commented:
> I would much rather we go down the path of building our own version. The fact that this PR is written from the perspective of an independent third-party conveying the wishes of "upstream", when the author is in fact upstream themself, feels malicious.

I'm not sure what exactly is malicious about this PR.  For the record, I am not the one making the final decisions there anymore, and haven't been for a long time, so when I imply that I am conveying the wishes of someone else as you say, that is in fact the truth. My interest here is that I run NixOS, and I have been using this package on my system for some time, and I want to make it available to others.

I am certainly not a 3rd party here though, and never intended to present myself that way. Anyone who even clicked on my name can clearly see that I'm associated with the upstream project.

jyooru commented on nixos/doc/manual/release-notes/rl-2205.section.md:
```diff
@@ -162,6 +164,11 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `idris2` now requires `--package` when using packages `contrib` and `network`, while previously these idris2 packages were automatically loaded.
 
+- `multimc` has been replaced with `multimc-bin`, which is a self-updating
+  package that uses official binaries from upstream. Users with existing
+  installations need not take any steps, as the new package stores data in the
+  same locations as the previous one.
+
```
```suggestion
- Due to upstream FOSS complications `multimc` has been moved to `multimc-bin`, which is a self-updating package that uses official, but unreproducible binaries. PolyMC is a fork that does not have MultiMC's FOSS issues and does not self update. Users with existing installations who want to switch to PolyMC should rename `~/.local/share/multimc` to `~/.local/share/polymc`. The main config file's path has also moved from `~/.local/share/multimc/multimc.cfg` to `~/.local/share/polymc/polymc.cfg`.
```

Forkk replied:
I think "upstream FOSS complications" seems a little vague. There should probably be some reasoning stated, but I'm not sure exactly what to say that would be appropriate for a changelog.

RaitoBezarius replied:
upstream FOSS complications could be a link to the discussion that led to remove MultiMC from nixpkgs and also this PR if needed at all.

jyooru commented on pkgs/games/multimc/icon.svg:
```diff
@@ -0,0 +1,353 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
It should not be necessary to store an icon in this package repository, especially if it is not considered FOSS. This should be fetched using a URL.

LunNova replied:
Should get squashed/rebased when removed so we don't accidentally include an unfree asset in the commit history.

Forkk replied:
I will update the package to download the icon instead of storing it in the repo. As for squashing, last time I contributed to nixpkgs (which was admittedly quite a long time ago), policy was to rebase on master and squash PRs into one commit before merging. Is that not still the case? If not, I will rebase such that the icon is not in the commit history.

LunNova replied:
I'm not sure what the rules are but my recent PRs weren't squashed when merged.

jyooru commented on pkgs/top-level/aliases.nix:
```diff
@@ -737,7 +737,6 @@ mapAliases ({
   mpd_clientlib = libmpdclient; # Added 2021-02-11
   mpich2 = mpich; # Added 2018-08-06
   msf = metasploit; # Added 2018-04-25
-  multimc = throw "multimc was removed from nixpkgs; use polymc instead (see https://github.com/NixOS/nixpkgs/pull/154051 for more information)"; # Added 2022-01-08
```
Should be restored. Could add a note to view the release notes instead of linking to the PR.

LunNova commented on pkgs/games/multimc/multimc.desktop:
```diff
@@ -0,0 +1,18 @@
+[Desktop Entry]
```
You can probably use `makeDesktopItem` for this instead.

https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/make-desktopitem/default.nix

Forkk committed:
multimc: set binary name back to `multimc`

This does not affect the package name, just the name of the binary in
the PATH.

Forkk committed:
multimc: use run command to generate symlinks

Forkk committed:
multimc: use makeDesktopItem

Forkk committed:
multimc: add back removal notice

flexagoon commented:
@Forkk what's the point of this when polymc is already packaged

Scrumplex commented:
Disclaimer: I am associated with the PolyMC project. I am only commenting the following for the sake of transparency, and why we chose to not hide our Microsoft client ID. This should not influence adoption of this package into nixpkgs.

@Forkk 

> and we are required by microsoft to keep the MSA ID needed for authentication as hidden as is practical

Microsoft's ToS explicitly differentiates between an `Application ID` and `Access Credentials`. The only thing that needs to be kept secret are those `Access Credentials`. In fact, to verify the use of a domain (domain as in DNS) in an Azure Active Directory application, you need to place said `Application ID` at a predictable (`.well-known`) directory on a web server. [[1]]


[1]: https://docs.microsoft.com/en-us/legal/microsoft-identity-platform/terms-of-use#3-application-registration

Forkk commented:
> @Forkk what's the point of this when polymc is already packaged

MultiMC and PolyMC are different projects with different goals. From what I can tell, PolyMC has plans to diverge from upstream MultiMC. This is like asking what the point of packaging vim is when neovim is already packaged.


> > and we are required by microsoft to keep the MSA ID needed for authentication as hidden as is practical
> 
> Microsoft's ToS explicitly differentiates between an `Application ID` and `Access Credentials`. The only thing that needs to be kept secret are those `Access Credentials`. In fact, to verify the use of a domain (domain as in DNS) in an Azure Active Directory application, you need to place said `Application ID` at a predictable (`.well-known`) directory on a web server. [[1](https://docs.microsoft.com/en-us/legal/microsoft-identity-platform/terms-of-use#3-application-registration)]

I'm not really interested in turning this into some argument over the MSA ID situation, and any such argument here would be unproductive anyway, as I don't contribute to MultiMC anymore aside from helping people in discord, and have basically no real decision making ability. I am here so I have something to point NixOS users at in our support channels. I don't care how that problem is solved, just that it is solved in a way we can support upstream.

I would appreciate if someone with commit access could weigh in on this, so I know whether it's worth investing more time into this PR, or if I should just turn it into a flake.

Sciencentistguy commented:
I would definitely be interested in having this as a flake / overlay fwiw.

rapenne-s commented:
In case this is going anywhere, the aliases multimc to polymc should be removed https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/aliases.nix#L908

RaitoBezarius commented:
I am not in favor of having this in nixpkgs, but I think it can be tolerated as `multimc-bin` with the changelog changes suggested by @jyooru, IMHO.

I think you will have a better time turning it into a Flake.
